File: packages/i18nify-js/src/modules/banking/__tests__/validatePAN.test.ts
import validatePAN from '../validatePAN';

describe('banking - validatePAN', () => {
  describe('valid PANs', () => {
    it('should accept a valid individual PAN (entity char P)', () => {
      expect(validatePAN('ABCPK1234L')).toBe(true);
    });

    it.each(['C', 'F', 'H', 'A', 'B', 'L', 'J', 'T', 'G'])(
      'should accept a valid PAN with entity char %s',
      (entity) => {
        expect(validatePAN(`ABC${entity}K1234L`)).toBe(true);
      },
    );

    it('should accept a lowercase PAN (case-insensitive)', () => {
      expect(validatePAN('abcpk1234l')).toBe(true);
    });

    it('should accept a PAN with surrounding whitespace', () => {
      expect(validatePAN('  ABCPK1234L  ')).toBe(true);
    });
  });

  describe('invalid PANs', () => {
    it('should reject an invalid 4th (entity) character', () => {
      // 'D' is not a recognised entity-type character.
      expect(validatePAN('ABCDE1234L')).toBe(false);
    });

    it('should reject a PAN that is too short', () => {
      expect(validatePAN('ABCPK123L')).toBe(false);
    });

    it('should reject a PAN that is too long', () => {
      expect(validatePAN('ABCPK1234LL')).toBe(false);
    });

    it('should reject a PAN with digits where letters are expected', () => {
      expect(validatePAN('AB1PK1234L')).toBe(false);
    });

    it('should reject a PAN with letters where digits are expected', () => {
      expect(validatePAN('ABCPKABCDL')).toBe(false);
    });

    it('should reject a PAN ending in a digit', () => {
      expect(validatePAN('ABCPK12345')).toBe(false);
    });

    it('should reject a PAN containing special characters', () => {
      expect(validatePAN('ABCPK-234L')).toBe(false);
    });
  });

  describe('invalid arguments', () => {
    it('should throw for an empty string', () => {
      expect(() => validatePAN('')).toThrow(/PAN/);
    });

    it('should throw for a non-string input', () => {
      // @ts-expect-error testing runtime guard against non-string input
      expect(() => validatePAN(null)).toThrow(/PAN/);
    });
  });
});

File: packages/i18nify-go/modules/banking/validate_pan_test.go
package banking

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// ── ValidatePAN ────────────────────────────────────────────────────────────────

func TestValidatePAN(t *testing.T) {
	tests := []struct {
		name    string
		pan     string
		want    bool
		wantErr bool
	}{
		{name: "valid individual PAN (entity P)", pan: "ABCPK1234L", want: true},
		{name: "valid company PAN (entity C)", pan: "ABCCK1234L", want: true},
		{name: "valid firm PAN (entity F)", pan: "ABCFK1234L", want: true},
		{name: "valid HUF PAN (entity H)", pan: "ABCHK1234L", want: true},
		{name: "valid trust PAN (entity T)", pan: "ABCTK1234L", want: true},
		{name: "valid govt PAN (entity G)", pan: "ABCGK1234L", want: true},
		{name: "lowercase PAN is accepted", pan: "abcpk1234l", want: true},
		{name: "surrounding whitespace is trimmed", pan: "  ABCPK1234L  ", want: true},

		{name: "invalid entity char D", pan: "ABCDE1234L", want: false},
		{name: "too short", pan: "ABCPK123L", want: false},
		{name: "too long", pan: "ABCPK1234LL", want: false},
		{name: "digit in alpha block", pan: "AB1PK1234L", want: false},
		{name: "letters in numeric block", pan: "ABCPKABCDL", want: false},
		{name: "ends with a digit", pan: "ABCPK12345", want: false},
		{name: "contains special character", pan: "ABCPK-234L", want: false},

		{name: "empty string returns error", pan: "", wantErr: true},
		{name: "whitespace-only returns error", pan: "   ", wantErr: true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidatePAN(tt.pan)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			assert.Equal(t, tt.want, got)
		})
	}
}
